state/execution.go

```diff
@@ -102,12 +104,88 @@ func (blockExec *BlockExecutor) CreateProposalBlock(
 
 	evidence, evSize := blockExec.evpool.PendingEvidence(state.ConsensusParams.Evidence.MaxBytes)
 
+	evdData := types.EvidenceData{Evidence: evidence}
+	pevdData, err := evdData.ToProto()
+	if err != nil {
+		// todo(evan): see if we can get rid of this panic
+		panic(err)
+	}
+
 	// Fetch a limited amount of valid txs
 	maxDataBytes := types.MaxDataBytes(maxBytes, evSize, state.Validators.Size())
 
+	// TODO(ismail): reaping the mempool has to happen in relation to a max
+	// allowed square size instead of (only) Gas / bytes
+	// maybe the mempool actually should track things separately
+	// meaning that CheckTx should already do the mapping:
+	// Tx -> Txs, Message
+	// https://github.com/tendermint/tendermint/issues/77
```

Comment:
This is an excellent point. I checked again the `PrepareProposal` implementation in Celestia-app ADR-006, and also in the Celestia-app [code](https://github.com/celestiaorg/celestia-app/blob/e088d61fcb6579b4bc797deefd2ceff7601aa079/app/prepare_proposal.go#L18). It is not taking into account `req.BlockDataSize` (`req` is the method's parameter).
I believe Celestia-core (Tendermint) should only care about block size, and Celestia-app should care about the encoding details.
If you agree with that, then probably, the way forward would be modify function `malleateTxs` so that it accepts `req.BlockDataSize` as a parameter and stops adding transactions to the output list whenever it detects that it has reached the block size limit.
This is consistent, by the way, with requirement 2 [here](https://github.com/tendermint/tendermint/blob/v0.37.x/spec/abci/abci%2B%2B_app_requirements.md), so we are expecting all applications to keep track of the total transaction size in `PrepareProposal`.

Reply:
As for `MaxGas`, you may want to deprecate it in Celestia-core.